File: CSharp/Legacy89DiskKit.FileSystem.Infrastructure/FileSystem/HuBasic/HuBasicNameRules.cs
namespace Legacy89DiskKit.FileSystem.Infrastructure.HuBasic;

public static class HuBasicNameRules
{
    public static (string FileName, string Extension) ParseFileName(string fileName)
    {
        if (fileName == null) throw new ArgumentNullException(nameof(fileName));

        var parts = fileName.Split('.');
        string name = parts[0];
        if (name.Length > 13)
        {
            name = name[..13];
        }

        string extension = parts.Length > 1 ? parts[1] : string.Empty;
        if (extension.Length > 3)
        {
            extension = extension[..3];
        }

        return (name, extension);
    }

    public static string BuildDisplayName(string fileName, string extension)
    {
        if (fileName == null) throw new ArgumentNullException(nameof(fileName));
        if (extension == null) throw new ArgumentNullException(nameof(extension));

        return string.IsNullOrEmpty(extension) ? fileName : $"{fileName}.{extension}";
    }
}

File: Cpp/Legacy89DiskKit.Cpp/src/hu_basic_name_rules.cpp
#include "legacy89diskkit/cpp/hu_basic_name_rules.hpp"

namespace legacy89diskkit::cpp
{
HuBasicParsedName HuBasicNameRules::ParseFileName(const std::string& file_name)
{
    const auto separator = file_name.find('.');
    auto name = separator == std::string::npos ? file_name : file_name.substr(0, separator);
    auto extension = separator == std::string::npos ? std::string() : file_name.substr(separator + 1);

    if (name.size() > 13)
    {
        name = name.substr(0, 13);
    }

    if (extension.size() > 3)
    {
        extension = extension.substr(0, 3);
    }

    return HuBasicParsedName{ name, extension };
}

std::string HuBasicNameRules::BuildDisplayName(const std::string& file_name, const std::string& extension)
{
    if (extension.empty())
    {
        return file_name;
    }

    return file_name + "." + extension;
}
}
